src/ERC1155D.sol

```diff
@@ -0,0 +1,388 @@
+        bytes memory data
+    ) internal virtual {
+        require(to != address(0), "ERC1155: transfer to the zero address");
+
+        address operator = msg.sender;
+        uint256[] memory ids = _asSingletonArray(id);
+        uint256[] memory amounts = _asSingletonArray(amount);
+
+        _beforeTokenTransfer(operator, from, to, ids, amounts, data);
+
+        require(ownerOf[id] == from && amount < 2, "ERC1155: insufficient balance for transfer");
+
+        // The ERC1155 spec allows for transfering zero tokens, but we are still expected
+        // to run the other checks and emit the event. But we don't want an ownership change
+        // in that case
+        if (amount == 1) {
+            ownerOf[id] = to;
+        }
+
+        emit TransferSingle(operator, from, to, id, amount);
+
+        _afterTokenTransfer(operator, from, to, ids, amounts, data);
+
+        _doSafeTransferAcceptanceCheck(operator, from, to, id, amount, data);
+    }
+
+    /**
+     * @dev xref:ROOT:erc1155.adoc#batch-operations[Batched] version of {_safeTransferFrom}.
+     *
+     * Emits a {TransferBatch} event.
+     *
+     * Requirements:
+     *
+     * - If `to` refers to a smart contract, it must implement {IERC1155Receiver-onERC1155BatchReceived} and return the
+     * acceptance magic value.
+     */
+    function _safeBatchTransferFrom(
+        address from,
+        address to,
+        uint256[] memory ids,
+        uint256[] memory amounts,
+        bytes memory data
+    ) internal virtual {
+        require(ids.length == amounts.length, "ERC1155: ids and amounts length mismatch");
+        require(to != address(0), "ERC1155: transfer to the zero address");
+
+        address operator = msg.sender;
+
+        _beforeTokenTransfer(operator, from, to, ids, amounts, data);
+
+        for (uint256 i = 0; i < ids.length; ++i) {
+            uint256 id = ids[i];
+
+            require(ownerOf[id] == from && amounts[i] < 2, "ERC1155: insufficient balance for transfer");
+
+            if (amounts[i] == 1) {
+                ownerOf[id] = to;
+            }
+        }
+
+        emit TransferBatch(operator, from, to, ids, amounts);
+
+        _afterTokenTransfer(operator, from, to, ids, amounts, data);
+
+        _doSafeBatchTransferAcceptanceCheck(operator, from, to, ids, amounts, data);
+    }
+
+    /**
+     * @dev Creates `amount` tokens of token type `id`, and assigns them to `to`.
+     *
+     * Emits a {TransferSingle} event.
+     *
+     * Requirements:
+     *
+     * - `to` cannot be the zero address.
+     * - `id` must be less than MAX_SUPPLY;
+     * This does not implement smart contract checks according to ERC1155 so it exists as a separate function
+     */
+
+    function _mintSingle(address to, uint256 id) internal virtual {
+        require(to != address(0), "ERC1155: mint to the zero address"); // you can remove this if only minting to msg.sender
+        require(ownerOf[id] == address(0), "ERC1155D: supply exceeded");
```

Comment:
Is this check necessary? As there is no supply limit, and duplicate checks if the token Id has already been minted.

Reply:
it's not. I am removing this in #65. My focus in the PR was to take 1155D without much changes and integrate it with Ditto. Basically minimizing the diff between this 1155D and the official 1155D.